source/blog/20019-01-22-sso-keycloak-ovirt-openidc-389ds.html.md

```diff
@@ -0,0 +1,267 @@
+
+READMORE
+
+## Keycloak
+Keycloak is an open source Identity and Access Management solution aimed at modern applications and services. It makes it easy to secure applications and services with little to no code.
+Users authenticate with Keycloak rather than individual applications. This means that your applications don't have to deal with login forms to authenticate users. Once logged in to Keycloak, users don't have to login again to access a different application.
+For our single sign on purpose we will use Keycloak as the OIDC provider.
+
+## Keycloak configuration
+For our example, we'll use default master realm. First let's create user federation with our `389ds` directory server.
+Fill the values of your 389ds server as show in the image below.
+
+![](../images/blog/2019-01-22/user_federation.png)
+
+Add ldap groups mapper
+
+![](../images/blog/2019-01-22/ldap_group_mapper.png)
+
+After testing the connection and authentication, click the `save` button And then Synchronize All Users using the button at the bottom of User Federation Provider page.. Check if your users are searchable, as shown in image below.
+![](../images/blog/2019-01-22/user_search.png)
+
+Now that user federation is working and is integrated, let's configure the OIDC client. 
+
+![](../images/blog/2019-01-22/add_client.png)
+
+**IMPORTANT** After the client is created, we must change `Access Token Lifespan` to `10 days`. This is required because oVirt manages the token lifespan and we do not want Keycloak invalidating the tokens.
+![](../images/blog/2019-01-22/life_span.png)
+
+Make a note of client id and client secret from the cleint credentials page, the client id and secret is needed to configure the apache mod_auth_openidc and also ovirt-engine.
+![](../images/blog/2019-01-22/client_id_secret.png)
+
+Create a mapper to add groups to the claim
+![](../images/blog/2019-01-22/groups_claim.png)
+
+Add the build in mapper for username
+![](../images/blog/2019-01-22/user_name.png)
+
+Create the scopes need by ovirt engine, ovirt-app-api and ovirt-app-admin
+![](../images/blog/2019-01-22/scope_ovirt_app_api.png)
+
+![](../images/blog/2019-01-22/scope_ovirt_app_admin.png)
+
+Set up optional client scopes for the ovirt-engine client
+![](../images/blog/2019-01-22/client_scopes.png)
+
+## mod_auth_openidc configuration
+First we need to configure oVirt's apache. SSH to the oVirt engine and install mod_auth_openidc and ovirt-engine-extension-aaa-misc.
+
+```bash
+ssh root@engineopenid
+yum install -y mod_auth_openidc
+yum install -y ovirt-engine-extension-aaa-misc
+```
+
+We need to create a file `/etc/httpd/conf.d/ovirt-openidc.conf` to configure mod_auth_openidc as follows:
+
+```
+LoadModule auth_openidc_module modules/mod_auth_openidc.so
+ServerName engineopenid.example.com
+
+OIDCProviderMetadataURL https://keycloak.example.com:8443/auth/realms/master/.well-known/openid-configuration
```

Comment:
Please mention somewhere at beginning that users need to replace "keycloak.example.com"  and all other example.com URLs with real URLs